### ISSUE TITLE
structure.column_order breaks subsequent UNION ALL.

### ISSUE THREAD

**Issue by rogalski** [sqlfluff/sqlfluff#5956]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

SQLFluff fix produces SQL with changed behavior.

### Expected Behaviour

SQL before and after fix is functionally equivalent.

### Observed Behaviour

T2 SELECT has reordered columns, which breaks UNION.

### How to reproduce

```sql
-- noqa: disable=structure.column_order
CREATE TABLE X AS
WITH T1 AS (
    SELECT
        'a'::varchar AS A,
        1::bigint AS B

),

T2 AS (
    SELECT
        CASE WHEN COL > 1 THEN 'x' ELSE 'y' END AS A,
        COL AS B
    FROM T
)

SELECT * FROM T1
UNION ALL
SELECT * FROM T2;
```

Run sqlfluff fix with disable statement, all good.
Run sqlfluff fix without disable statement, output columns in T2 are reordered, which breaks subsequent union.

### Dialect

snowflake

### Version

3.0.7

### Configuration

default

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by sbrugman** [comment]
Encountered the same issue with `sparksql`, resulting in:

> pyspark.sql.utils.AnalysisException: Union can only be performed on tables with the compatible column types.

**Comment by sbrugman** [comment]
Source that needs to be patched:
https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/rules/structure/ST06.py
